### PR TITLE
Implement efficient PersistentVectorBuilder iterator

### PR DESCRIPTION
**Pull request by qurbonzoda**

### CODE REVIEW

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt [hunk 1]
```diff
@@ -26,12 +26,14 @@ private class Marker // TODO: Rename to MutabilityOwnership?
 class PersistentVectorBuilder<E>(private var vector: PersistentList<E>,
                                  private var vectorRoot: Array<Any?>?,
                                  private var vectorTail: Array<Any?>,
-                                 private var rootShift: Int) : AbstractMutableList<E>(), PersistentList.Builder<E> {
+                                 internal var rootShift: Int) : AbstractMutableList<E>(), PersistentList.Builder<E> {
     private var marker = Marker()
-    private var root = vectorRoot
-    private var tail = vectorTail
+    internal var root = vectorRoot
+    internal var tail = vectorTail
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt]: If these are only read from outside makes sense to limit visibility of _setter_ to private

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        return trieIterator.previous()
+    }
+
+    override fun next(): T {
+        checkForComodification()
+        checkHasNext()
+
+        lastIteratedIndex = index
+
+        val trieIterator = this.trieIterator ?: return builder.tail[index++] as T
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt]: How comes `tail[index]` doesn't cause index-out-of-bound exception?

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            return
+        }
+
+        val trieSize = rootSize(builder.size)
+        val trieIndex = index.coerceAtMost(trieSize)
+        val trieHeight = builder.rootShift / LOG_MAX_BUFFER_SIZE + 1
+        if (trieIterator == null) {
+            trieIterator = TrieIterator(root, trieIndex, trieSize, trieHeight)
+        } else {
+            trieIterator?.reset(root, trieIndex, trieSize, trieHeight)
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt]: Better to use `!!` if `trieIterator` is expected to be not null here.

Comment (ilya-g) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt]: Also can be written without null assertion as 
```
trieIterator?.apply { reset(...) } ?: 
    run { trieIterator = TrieIterator(...) }
```

File: kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/stressTests/immutableList/PersistentListBuilderTest.kt [hunk 4]
```diff
@@ -286,6 +287,180 @@ class PersistentListBuilderTest {
         }
     }
 
+    private fun <E> testIteratorsState(expected: ListIterator<E>, actual: ListIterator<E>) {
+        assertEquals(expected.hasPrevious(), actual.hasPrevious())
+        assertEquals(expected.previousIndex(), actual.previousIndex())
+        assertEquals(expected.hasNext(), actual.hasNext())
+        assertEquals(expected.nextIndex(), actual.nextIndex())
+    }
```
Comment (ilya-g) [inline kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/stressTests/immutableList/PersistentListBuilderTest.kt]: Looks similar to `compare(expected, actual) { propertyEquals { ... } }` from other collection tests

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt (hunk 2 above)
Comment (ilya-g) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt]: Ok, I've got that if `trieIterator` is null the list consists of a tail only and index is actually the index in tail. Perhaps worth to document somewhere, as this is unobvious from at first sight.

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt (hunk 1 above)
Comment (qurbonzoda) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt]: +1

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt (hunk 2 above)
Comment (qurbonzoda) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt]: +1

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt (hunk 3 above)
Comment (qurbonzoda) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorMutableIterator.kt]: Used !! operator as it seem more readable.

File: kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/stressTests/immutableList/PersistentListBuilderTest.kt (hunk 4 above)
Comment (qurbonzoda) [reply kotlinx-collections-immutable/tests/src/test/kotlin/kotlinx.collections.immutable/stressTests/immutableList/PersistentListBuilderTest.kt]: It turns out `compare(expected, actual) { listIteratorProperties() }` does exactly what is needed.

**Comment by ilya-g** [comment]
Is there a test for comodification detection?

**Comment by qurbonzoda** [comment]
Test for comodification are added here: https://github.com/Kotlin/kotlinx.collections.immutable/pull/33

File: kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt [hunk 5]
```diff
@@ -26,12 +26,16 @@ private class Marker // TODO: Rename to MutabilityOwnership?
...
-                                 private var rootShift: Int) : AbstractMutableList<E>(), PersistentList.Builder<E> {
+                                 internal var rootShift: Int) : AbstractMutableList<E>(), PersistentList.Builder<E> {
     private var marker = Marker()
-    private var root = vectorRoot
-    private var tail = vectorTail
+    internal var root = vectorRoot
+        private set
+    internal var tail = vectorTail
+        private set
     override var size = vector.size
```
Comment (ilya-g) [inline kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt]: This doesn't seem to be changed from outside too

Comment (qurbonzoda) [reply kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVectorBuilder.kt]: +1

**Pull request closed without merging**